src/utils/hotcode-agent/src/main/java/io/bellsoft/hotcode/agent/HotCodeAgent.java

```diff
@@ -66,18 +84,14 @@ private static void parseArgs(String argumentString, Properties properties) {
     }
 
     private class ProfileAndOptimize implements Runnable {
-        private ProfilingTask profilingTask;
-
-        public ProfileAndOptimize(ProfilingTask profilingTask) {
-            this.profilingTask = profilingTask;
-        }
-
         @Override
         public void run() {
             try {
-                var hotMethods = profilingTask.call().getTop();
+                profiling.fill(profile);
+                var hotMethods = profile.getTop(config.top);
                 var directives = CompilerDirectives.build(hotMethods);
-                directivesControl.replace(directives);
+                CompilerDirectivesControl.replace(directives);
+                profile.clear();
```

Comment:
Should the agent log something about successful profiling and application of directives?
I think it also makes sense to report some stats through MXBean so that the process can be monitored externally.